w3af/plugins/audit/hpp.py

```diff
@@ -0,0 +1,263 @@
+    result = []
+    #find reflection in links
+    for references in parser.get_references():
+        for ref in references:
+            if not ref:
+                continue
+            qs = ref.get_querystring()
+            for p in qs:
+                if precedence == LAST and len(qs[p]) == 2:
+                    rv = qs[p][1]
+                else:
+                    rv = qs[p][0]
+                #looking for single reflection if only one pattern was
+                # entered
+                if rv == pattern1 and not pattern2:
+                    result += [(p, ref)]
+
+                if not pattern2:
+                    continue
+                #or for concatenation of both were
+                if rv[:len(pattern1)] == pattern1 and \
+                   rv[-len(pattern2):] == pattern2:
+                    result += [p, ref]
+
+    return result
+
+
+class hpp(AuditPlugin):
+    """
+    Identify client-side http parameter pollution vulnerabilities.
+    """
+
+    def __init__(self):
+        AuditPlugin.__init__(self)
+        self.parser = etree.HTMLParser(remove_blank_text=True, remove_comments=True)
+
+    def audit(self, freq, orig_response):
+        """
+        Tests an URL for HPP vulnerabilities.
+        :author: Dmitry Roshchin (nixwizard@gmail.com)
+        """
+
+        if not orig_response.is_text_or_html():
+            return
+
+        if not orig_response.get_code() == 200:
+            return
+
+        method = freq.get_method()
+
+        if method == 'GET':
+            qs = freq.get_querystring()
+
+        precedence = None
+        #finding reflected params if any
+        for param in qs:
+            value = qs[param][0]
+            #checking if the value is number or string so that our
+            #reflection test request would be more legitimate
+            if value.isdigit():
+                newval = unicode(int(value)+1)
+            else:
+                newval = REFLECT_TOKEN
+
+            if precedence:
+                continue
+
+            mutants = create_mutants(freq, [newval],
+                                     fuzzable_param_list=[param],
+                                     mutant_tuple=(QSMutant,))
+            mu = mutants[0]
+            #doing reflection test for the param
+            rtest_resp = self._uri_opener.send_mutant(mu)
+            if not rtest_resp.get_code() == 200:
+                #no need to seek reflection if it is not OK
+                continue
+
+            if orig_response.body == rtest_resp.body:
+                #parameter param is not being reflected going for the next
+                continue
+
+            #Now we are doing precedence test
+            u = str(mu.get_uri())
+            #this hack is to make w3af send request with unencoded qs
+            #because as long as we are trying to use qs it urlencodes the
+            #second instance of param with new value
+            ptest_url = URL(u.replace(newval,
```

Comment:
What if you're testing `http://foo122.com/?id=121` ?

That would replace the domain too!